### PR TITLE
Remove `ARFLAGS` hack for Windows, replace with `TEMPFILE`

### PR DESCRIPTION
**Pull request by alvinhochun**
TEMPFILE is the built-in way of SCons to use a response file for command lines that are too long.

@bruvzg This can replace #96220.

### CODE REVIEW

**Review by bruvzg — APPROVED** [review]

File: platform/windows/detect.py [hunk 1]
```diff
@@ -654,6 +654,11 @@ def configure_mingw(env: "SConsEnvironment"):
     # https://www.scons.org/wiki/LongCmdLinesOnWin32
     env.use_windows_spawn_fix()
 
+    # In case the command line to AR is too long, use a response file.
+    env["ARCOM_ORIG"] = env["ARCOM"]
+    env["ARCOM"] = "${TEMPFILE('$ARCOM_ORIG', '$ARCOMSTR')}"
+    env["TEMPFILESUFFIX"] = ".rsp"
```
Comment (Faless) [inline platform/windows/detect.py:660]: If we want to follow SCons, we might want to do like done in their [msvc tool](https://github.com/SCons/scons/blob/master/SCons/Tool/mslib.py#L60).

For msvc it should be:
```
env['ARCOM'] = "${TEMPFILE('$AR $ARFLAGS /OUT:$TARGET $SOURCES','$ARCOMSTR')}"
```

For [all others](https://github.com/SCons/scons/blob/master/SCons/Tool/ar.py#L47) (including mingw), I guess it should be:

```
env['ARCOM'] = "${TEMPFILE('$AR $ARFLAGS $TARGET $SOURCES','$ARCOMSTR')}"
```

Comment (alvinhochun) [reply platform/windows/detect.py:660]: That was what I originally had, but I didn't want to duplicate what had been defined in SCons itself. The approach I ended up with is similar to what the web platform is using: https://github.com/godotengine/godot/blob/7c383767a29969b0c3db036742da00ef58b765dd/platform/web/detect.py#L184-L185

Comment (Faless) [reply platform/windows/detect.py:660]: I see, I feel that also should also be changed to be honest.

That risks creating unexpected chains of TEMPFILE if the setup function is called multiple times.

If we know the toolchain, we also know the `ARCOM` format for that toolchain, so we should enforce it I think.

Comment (alvinhochun) [reply platform/windows/detect.py:660]: Nesting `TEMPFILE` doesn't affect execution - `TempFileMunge` only creates and uses the temp file if the actual command line exceeds the set length, so only the deepest level of `TEMPFILE` takes effect, while outer ones see the already shortened command line. (I have tested it.)

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/02b16d2f544e323b7b7f57e6e992b0b8e5d8b954), it works as expected on Windows 11 23H2 + MSVC 2022 (17.10.0).

**Review by Faless — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
This seems to have broken mingw build on Windows (at least with some mingw distribution). I saw this reported on Discord by @noidexe:
```
[Initial build] Linking Static Library modules\freetype\libfreetype_builtin.windows.editor.x86_64.a ...
ERROR: C:/Users/lisan/scoop/apps/mingw/14.2.0-rt_v12-rev0/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ar.exe: thirdpartyfreetypesrcautofitautofit.windows.editor.x86_64.o: No such file or directory

scons: *** [modules\freetype\libfreetype_builtin.windows.editor.x86_64.a] Error 1
scons: building terminated because of errors.
```

**Comment by AThousandShips** [comment]
Was just about to post on this but I get a different error, involving prefixed `@` on tempfiles:
```
x86_64-w64-mingw32-gcc-ar @C:\Users\XXXX\AppData\Local\Temp\tmpzcvaw526.rsp
x86_64-w64-mingw32/bin/ar.exe: unknown option -- @
```

**Comment by alvinhochun** [comment]
> This seems to have broken mingw build on Windows (at least with some mingw distribution). I saw this reported on Discord by @noidexe:
> 
> ```
> [Initial build] Linking Static Library modules\freetype\libfreetype_builtin.windows.editor.x86_64.a ...
> ERROR: C:/Users/lisan/scoop/apps/mingw/14.2.0-rt_v12-rev0/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ar.exe: thirdpartyfreetypesrcautofitautofit.windows.editor.x86_64.o: No such file or directory
> 
> scons: *** [modules\freetype\libfreetype_builtin.windows.editor.x86_64.a] Error 1
> scons: building terminated because of errors.
> ```

I see, so it appears we will have to set [`TEMPFILEARGESCFUNC`](https://scons.org/doc/production/HTML/scons-user/apa.html#cv-TEMPFILEARGESCFUNC) to replace backslashes to forward slashes.



> Was just about to post on this but I get a different error, involving prefixed `@` on tempfiles:
> 
> ```
> x86_64-w64-mingw32-gcc-ar @C:\Users\XXXX\AppData\Local\Temp\tmpzcvaw526.rsp
> C:/Users/XXX/scoop/apps/mingw/13.2.0-rt_v11-rev0/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ar.exe: unknown option -- @
> ```

Hmm, Binutils ar should support response file. This doesn't look right.

**Comment by AThousandShips** [comment]
My above issue was solved by updating to mingw 14.2, unsure when the relevant functionality was introduced or exactly what was different but might be worth looking at versions supporting this

**Comment by Repiteo** [comment]
Can confirm this caused me issues with adding GCC to the Windows CI in #97446; reverting the change allowed the build to work as expected

**Comment by alvinhochun** [comment]
> Was just about to post on this but I get a different error, involving prefixed `@` on tempfiles:
> 
> ```
> x86_64-w64-mingw32-gcc-ar @C:\Users\XXXX\AppData\Local\Temp\tmpzcvaw526.rsp
> C:/Users/XXX/scoop/apps/mingw/13.2.0-rt_v11-rev0/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ar.exe: unknown option -- @
> ```
> 
> This was resolved by updating to mingw 14.2



> Can confirm this caused me issues with adding GCC to the Windows CI in #97446; reverting the change allowed the build to work as expected

This took me a while to diagnose, but turns out it was a known upstream GCC bug with calling `gcc-ar` (instead of plain `ar`) until quite recently, which inserts a dash in front of the first argument even if it specifies a response file: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77576
Looks like the fix landed on GCC 14.1.0.

Some ideas to accommodate older versions:
- Don't use `gcc-ar`, just use `ar` directly?
    - From the comments in the bug report it seems `ar` should be able to find the LTO plugin automatically, so `gcc-ar` should not be necessary these days. This needs testing.
- Override `TempFileMunge` to leave the ARFLAGS outside of the response file (i.e. the command becomes `gcc-ar.exe rc --thin @C:\temp\tmp123.rsp`)
    - Seems non-trivial
- Add a hack to `use_windows_spawn_fix` to insert `-- ` before the response file command line argument (i.e. the command becomes `gcc-ar.exe -- @C:\temp\tmp123.rsp`)
    - Does not fix cross-compilation from Linux, though with the long command line limit on Linux it is less of an issue there.

The first option may be the best if it works. Can someone test it out? (I very much doubt my system can handle an LTO build of Godot.)

**Comment by AThousandShips** [comment]
If you have some diffs for the suggested changes I can try them out

**Comment by alvinhochun** [comment]
> If you have some diffs for the suggested changes I can try them out

Probably like

```diff
diff --git a/platform/windows/detect.py b/platform/windows/detect.py
index db4c743595..7fbcd8a0df 100644
--- a/platform/windows/detect.py
+++ b/platform/windows/detect.py
@@ -792,8 +792,6 @@ def configure_mingw(env: "SConsEnvironment"):
         env["CXX"] = mingw_bin_prefix + "g++"
         if try_cmd("as --version", env["mingw_prefix"], env["arch"]):
             env["AS"] = mingw_bin_prefix + "as"
-        if try_cmd("gcc-ar --version", env["mingw_prefix"], env["arch"]):
-            env["AR"] = mingw_bin_prefix + "gcc-ar"
         if try_cmd("gcc-ranlib --version", env["mingw_prefix"], env["arch"]):
             env["RANLIB"] = mingw_bin_prefix + "gcc-ranlib"
```

(I do also wonder if `gcc-ranlib` is needed in a similar sense...)

**Comment by Repiteo** [comment]
I can confirm that commenting out the `gcc-ar` code fixed the issue for my PR!

**Comment by Faless** [comment]
> I can confirm that commenting out the `gcc-ar` code fixed the issue for my PR!

Doesn't that risk breaking creoss-compilation from other platforms though?

**Pull request merged**